File: server/models/newsletter_subscription.py
from pydantic import BaseModel, EmailStr, Field
from typing import Optional, List
from datetime import datetime

class NewsletterSubscription(BaseModel):
    id: Optional[str]  # MongoDB ObjectId
    email: EmailStr
    status: str = Field(default="subscribed", regex="^(subscribed|unsubscribed|bounced)$")
    verification_token: Optional[str]
    verified: bool = False
    verified_at: Optional[datetime]
    unsubscribed_at: Optional[datetime]
    last_email_sent_at: Optional[datetime]
    tags: List[str] = []
    ip_address: Optional[str]
    user_agent: Optional[str]
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

File: server/models/NewsletterSubscription.js
const mongoose = require('mongoose');

const newsletterSubscriptionSchema = new mongoose.Schema({
    email: {
        type: String,
        required: [true, 'Email is required'],
        unique: true,
        lowercase: true,
        trim: true,
        match: [/^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/, 'Please provide a valid email']
    },
    status: {
        type: String,
        enum: ['subscribed', 'unsubscribed', 'bounced'],
        default: 'subscribed'
    },
    verificationToken: String,
    verified: {
        type: Boolean,
        default: false
    },
    verifiedAt: Date,
    unsubscribedAt: Date,
    lastEmailSentAt: Date,
    tags: [String], // For segmentation
    ipAddress: String,
    userAgent: String,
}, { timestamps: true });

// Index for search (email has unique constraint, createdAt indexed via timestamps option)
newsletterSubscriptionSchema.index({ status: 1 });

const NewsletterSubscription = mongoose.model('NewsletterSubscription', newsletterSubscriptionSchema);
module.exports = NewsletterSubscription;

